frontend/apps/reader/modules/readeruserhyph.lua

```diff
@@ -0,0 +1,372 @@
+    end
+    return false
+end
+
+-- this formats a suggestion with upper/lowercase as given in word
+-- Example suggestion=buß-jäger; word=BUẞJäger
+-- return: BUẞ-Jäger
+-- notice the 'ß' is coded in two bytes, whereas 'ẞ' uses three bytes. (terrible mapping)
+function ReaderUserHyph:formatHyphenation(suggestion, word)
+    if not suggestion then return word end
+
+    local suggestion_lower = Utf8:lower(suggestion)
+    local word_lower = Utf8:lower(word)
+    local val = ""
+    local i, j = 1, 1
+    local word_len = word_lower:len()
+    while i <= word_len do
+        if suggestion:sub(j,j) == "-" then
+            val = val .. "-"
+            j = j + 1
+        end
+        if word_lower:sub(i, i) == suggestion_lower:sub(j, j) then
+            -- e.g.: lowercase ß (Utf8: 0xC3 0x9F), uppercase ẞ (Utf8: 0xE1 0xBA 0x9E)
+            -- so a matching lowercase must be translated to uppercase if necessary
+            if word_lower:sub(i, i) == word:sub(i,i) then -- word[i] is lowercase, then dumb copy
+                val = val .. word:sub(i, i) -- use uppercases from word, as suggestion is always lowercase
+                i = i + 1
+                j = j + 1
+            else  -- case differs in word[i] and word_lower[i]
+                local multibyte = false
+                for low_c, up_c in pairs(Utf8.lc_uc) do
+                    if (word_lower:sub(i):find(low_c) or 0) == 1 then -- Utf8-multibyte at position i
+                        print("multibyte ".. low_c .. " " .. up_c)
+                        val = val .. up_c
+                        local len = low_c:len()
+                        i = i + len -- plus 1, 2 or 3; multibyte count
+                        j = j + len -- plus 1, 2 or 3; multibyte count
+                        multibyte = true
+                        break
+                    end
+                end
+                if not multibyte and i <= word_len then
+                    val = val .. word:sub(i, i) -- should not happen; if so, dumb copy
+                    i = i + 1
+                    j = j + 1
+                end
+            end
+
+        end
+    end
+    return val
+end
+
+-------------------------------------
+-- dictionary, file functions use:
+--    self.dict_file
+--    self.new_dict_file
+--    self.dict
+--    self.next_line
+-------------------------------------
+
+-- opens user dictionary
+function ReaderUserHyph:openDictionary()
+    self.dict_file = self.ui.userhyph:getDictionaryPath()
+    self.new_dict_file = self.dict_file .. ".new"
+
+    -- if no file, crate an empty one
+    if lfs.attributes(self.dict_file, "mode") ~= "file" then
+        self.dict = io.open(self.dict_file, "w")
+        self.dict:close()
+    end
+
+    -- check user dict file
+    if string.sub(lfs.attributes(self.dict_file).permissions, 1, 2) ~= "rw" then
+        UIManager:show(InfoMessage:new{
+            text = T(_("Wrong file permissions for:\n\"%1\""), self.dict_file),
```

Comment:
Do we do permissions check any place else ?

Reply:
I have not checked it. 
But I think it is sane to do it here, otherwise the error handling would be much more difficult.

Reply:
Well, if we don't any of these checks elsewhere, we assume everything we touch is usable (we're on single user devices, on FAT partitions).
So, no error handling needed. No need to add too much and make the code larger.

Reply:
That seems more like something to do in this general area, should we want to do it:

https://github.com/koreader/koreader/blob/b938f312a6245f25a2e9713c7da4787603973c43/reader.lua#L273-L277

It could be helpful in some edge cases on the desktop or something, but not on a device.